Previews are broken

@shiftingsand let me know about:

"FYI, you can’t view previews for *ANY* chart because of this error. "

![IMG_7278](https://user-images.githubusercontent.com/170948/178123963-d083c674-09b4-479a-bcab-c89a60024447.jpeg)
![IMG_3445](https://user-images.githubusercontent.com/170948/178123964-19a01f08-cf23-469a-9adf-cadf452a5e99.jpeg)

FYI, other charts (such as SoundBars) look like they have the same problem.  Preview won't show.

ambiguous use of operator '-'

----------------------------------------

CompileDylibError: Failed to build SoundBars.swift

Compiling failed: ambiguous use of operator '-'

Swift-Charts-Examples/Swift Charts Examples/Charts/BarCharts/SoundBars.swift:71:30: error: ambiguous use of operator '-'
        .chartYScale(domain: -__designTimeInteger("#8148.[2].[4].property.[0].[0].modifier[2].arg[0].value.[0]", fallback: 1)...__designTimeInteger("#8148.[2].[4].property.[0].[0].modifier[2].arg[0].value.[1]", fallback: 1))
                             ^
Swift.Float16:3:31: note: found this candidate
    prefix public static func - (x: Float16) -> Float16
                              ^
Swift.Float:2:31: note: found this candidate
    prefix public static func - (x: Float) -> Float
                              ^
Swift.Double:2:31: note: found this candidate
    prefix public static func - (x: Double) -> Double
                              ^

> FYI, other charts (such as SoundBars) look like they have the same problem. Preview won't show.
> 
> ambiguous use of operator '-'
> 
> CompileDylibError: Failed to build SoundBars.swift
> 
> Compiling failed: ambiguous use of operator '-'
> 
> Swift-Charts-Examples/Swift Charts Examples/Charts/BarCharts/SoundBars.swift:71:30: error: ambiguous use of operator '-' .chartYScale(domain: -__designTimeInteger("#8148.[2].[4].property.[0].[0].modifier[2].arg[0].value.[0]", fallback: 1)...__designTimeInteger("#8148.[2].[4].property.[0].[0].modifier[2].arg[0].value.[1]", fallback: 1)) ^ Swift.Float16:3:31: note: found this candidate prefix public static func - (x: Float16) -> Float16 ^ Swift.Float:2:31: note: found this candidate prefix public static func - (x: Float) -> Float ^ Swift.Double:2:31: note: found this candidate prefix public static func - (x: Double) -> Double ^

Soundbars previews fixed via https://github.com/jordibruin/Swift-Charts-Examples/commit/ef801a6958ba6ce413ae7b4a86094676c742a7e4

--- a/Swift Charts Examples/Charts/BarCharts/PyramidChart.swift
+++ b/Swift Charts Examples/Charts/BarCharts/PyramidChart.swift
@@ -43,17 +43,7 @@ struct PyramidChart: View {
 	private var chart: some View {
 		Chart(data) { series in
 			ForEach(series.population, id: \.percentage) { element in
-                Plot {
-                    BarMark(
-                        xStart: .value("Percentage", 0),
-                        xEnd: .value("Percentage", series.sex == "Male" ? element.percentage : element.percentage * -1),
-                        y: .value("AgeRange", element.ageRange),
-                        height: isOverview ? .automatic : .fixed(barHeight)
-                    )
-                }
-                .accessibilityLabel("\(series.sex); Ages: \(description(for: element.ageRange)) years")
-                .accessibilityValue("\(element.percentage.formatted(.percent))")
-                .accessibilityHidden(isOverview)
+                chartContent(series: series, element: element)
 			}
 			.foregroundStyle(series.sex == "Male" ? rightColor.gradient : leftColor.gradient)
 			.interpolationMethod(.catmullRom)
@@ -79,6 +69,22 @@ struct PyramidChart: View {
 		.chartXAxis(isOverview ? .hidden : .automatic)
 		.frame(height: isOverview ? 200 : 340)
 	}
+
+    // Workaround for previews. Swift compiler seems to have problems checking expression type
+    // for this chart when used directly inside Chart/ForEach, and the preview builds end up timing out.
+    @ChartContentBuilder func chartContent(series: PopulationByAgeData.Series, element: PopulationByAgeData.Series.Population) -> some ChartContent {
+        Plot {
+            BarMark(
+                xStart: .value("Percentage", 0),
+                xEnd: .value("Percentage", series.sex == "Male" ? element.percentage : element.percentage * -1),
+                y: .value("AgeRange", element.ageRange),
+                height: isOverview ? .automatic : .fixed(barHeight)
+            )
+        }
+        .accessibilityLabel("\(series.sex); Ages: \(description(for: element.ageRange)) years")
+        .accessibilityValue("\(element.percentage.formatted(.percent))")
+        .accessibilityHidden(isOverview)
+    }
     
     private var customisation: some View {
         Group {

--- a/Swift Charts Examples/ContentView.swift
+++ b/Swift Charts Examples/ContentView.swift
@@ -5,31 +5,10 @@
 import SwiftUI
 
 struct ContentView: View {
+    @StateObject private var imageCache = ImagesCache()
     @State private var selectedChartType: ChartType?
     @State var filterCategory: ChartCategory = .all
-    
-    private var cachedChartImages: [String: UIImage] = [:]
-    
-    @MainActor
-    init() {
-        ChartType.allCases.forEach { chart in
-            let view = chart.view
-                .padding(10)
-                .frame(width: 300)
-                .background {
-                    if chart.category == .apple {
-                        RoundedRectangle(cornerRadius: 10)
-                            .foregroundColor(.white)
-                    }
-                }
-            let renderer = ImageRenderer(content: view)
-            renderer.scale = UIScreen.main.scale
-            if let image = renderer.uiImage {
-                cachedChartImages[chart.id] = image
-            }
-        }
-    }
-    
+
     var body: some View {
         NavigationSplitView {
             List(selection: $selectedChartType) {
@@ -98,11 +77,34 @@ struct ContentView: View {
                 
                 // workaround to address hanging UI
                 // Reported FB10335209
-                if let image = cachedChartImages[chart.id] {
+                if let image = imageCache.images[chart] {
                     AccessiblePreviewImage(id: chart.id, image: image)
                 }
             }
         )
+    }
+}
+
+@MainActor private final class ImagesCache: ObservableObject {
+    @Published var images: [ChartType: UIImage] = [:]
+
+    init() {
+        ChartType.allCases.forEach { chart in
+            let view = chart.view
+                .padding(10)
+                .frame(width: 300)
+                .background {
+                    if chart.category == .apple {
+                        RoundedRectangle(cornerRadius: 10)
+                            .foregroundColor(.white)
+                    }
+                }
+            let renderer = ImageRenderer(content: view)
+            renderer.scale = UIScreen.main.scale
+            if let image = renderer.uiImage {
+                images[chart] = image
+            }
+        }
     }
 }
 

--- a/Swift Charts Examples/Data/Data.swift
+++ b/Swift Charts Examples/Data/Data.swift
@@ -234,11 +234,13 @@ enum LocationData {
 struct PopulationByAgeData {
     /// A data series for the bars.
     struct Series: Identifiable {
+        typealias Population = (ageRange: String, percentage: Int)
+
         /// Sex.
         let sex: String
 
         /// Percentage population for ageRange
-        let population: [(ageRange: String, percentage: Int)]
+        let population: [Population]
 
         /// The identifier for the series.
         var id: String { sex }

